File: MDScpp/MDScpp/IDC.cpp
#include "IDC.h"
#include "IDCDimension.h"
#include "IStrategie.h"
#include "LPS.h"
#include "SRPS.h"
#include "VariableLPS.h"
#include "VariableSRPS.h"

IDC::IDC(const vector<int> &minCell, const vector<int> &maxCell) : 
		initPoint(minCell), dimensionality(minCell.size()), shape(dimensionality,0),
		dimensions(dimensionality), positions(dimensionality,0), werte(0), realSize(0),
		precomputed(false), totalSum(0), min_array(new double[dimensionality]), max_array(new double[dimensionality])
		 {
	if (rangeValidator(minCell, maxCell))
		{
		// Shape wird gefuellt
		for (int i=0; i<dimensionality;i++)
			{
			shape[i]=maxCell[i]-minCell[i]+1;
			}
		// Dimensions wird gefuellt
		for (int i = 0; i < dimensionality; i++) {
			dimensions[i]=IDCDimension(shape[i]);
			}
		int tmp = 1;
		// Positions wird gefuelt
		positions[dimensionality - 1] = 1;
		for (int i = 2; i <= dimensionality; i++) {
			tmp *= shape[dimensionality - i + 1];
			positions[dimensionality - i] = tmp;
			}
		// Werte Array wird auf sine Kapazität vergroßert
		werte.resize(tmp*(shape[0]),0);

		for(int i=0; i<dimensionality; i++) {
			min_array[i]=minCell[i];
			max_array[i]=maxCell[i];
			}

		//cout<<"IDC Konstruktor wurde aufgerufen.\n";
		} 
	else
		{
		cerr<<"IDC koennte nicht gebaut werden!\n";
		}	
	}

IDC::~IDC() 
	{
		//cout<<"IDC Destruktor wurde aufgerufen!\n";
	}

void IDC::setCellValue( vector<int> &coordinates, double value) {

	for(int i=0; i< coordinates.size();i++)
		{
		coordinates[i] -= initPoint[i];
		}

		if (coordinatesValidation(coordinates)) {
		int pos = getPosition(coordinates);
		if (!precomputed) {
			if (value == 0) {
				if (werte[pos] != 0) {
					realSize--; // Wert wird auf 0 gesetzt
					totalSum=totalSum-value;
					}
				} 
			else if (werte[pos] == 0) {
				realSize++;
				totalSum=totalSum+value;
				}
			werte[pos] = value;
			} else {
				rangeUpdate(coordinates, coordinates, value, 0);
			}
		}
	}


void IDC::precompute(){
	 if (precomputed) {
		    //cout<<"It is precomuted!";
            return;
        }
	 else{
        for (int i = 0; i < dimensionality; i++) {
			IStrategie *tmp = dimensions[i].getStrategie();
            //Standardstrategie setzen
            if (tmp == 0) {
				int isize(dimensions[i].getSize());
				double dsize(isize);
				//tmp = new LPS( isize, static_cast<int>(sqrt(dsize)/1));
				tmp = new LPS( isize, dsize);
				dimensions[i].setStrategie(tmp);
            }
			tmp->precompute(this,i);

        }
		precomputed = true;
		//cout<<"Precompute is executed!\n";
	    }
	}


// TODO: Komentare durcharbeiten

double IDC::getRangeSum(const vector<int> &lower, const vector<int> &upper) {
	if(!rangeValidator(lower,upper) && !precomputed){
		return -1;
		}
	else {
		//Speichert das Zwischenergebnis
		double sum=0;
		//Enthaelt die Indizes fuer alle Dimensionen
		vector<int> indizies(0);
		//vector<int> pos enthaelt die erste Position, die einen Index fuer die Dimension i enthaelt
		vector<int> pos(dimensionality);
		//Hilfsvariable zur Berechnung des korrekten pos-Arrays
		int position = 0;
		//Speichert die Anzahl an Werten, die man zur Summe hinzufuegen muss
		int totalNumber = 1;
		//Fuegt die Indizes aller Dimensionen zu einer Liste zusammen
		for (int i = 0; i < dimensionality; i++ ) {
			vector<int> tmp( dimensions[i].getStrategie()->getCellsForRangeSum(lower[i], upper[i]));
			indizies.insert(indizies.end(),tmp.begin(), tmp.end());
			pos[i] = tmp.size();
			totalNumber *= pos[i];
			position += pos[i];
			}
		//Enthaelt fuer jede Dimension Anzahl der Indizes
		vector<int> shape(pos.begin(),pos.end());
		//Berechnet korrektes Indizes Array
		for (int i=dimensionality-1; i>=0; i--) {
			pos[i]=position - pos[i];
			position = pos[i];
			}
		//Hilfsarray, das dafuer sorgt, dass alle moeglichen Indexkombinationen durchlaufen werden.
		vector<int> current(dimensionality);
		//enthaelt die Positionen in der Liste, wo die Koordinaten drinstehen
		vector<int> ListPos(0);

		for(int i=0; i < totalNumber; i++) {
			ListPos = getListPositions(current,pos);
			//enthaelt die Koordinaten, die zur Summe dazugerechnet werden muessen
			vector<int> coordinates(dimensionality);
			//Vorzeichen des Summanden
			int vz = 1;
			for(int k=0; k<dimensionality; k++) {
				//liest Koordinaten aus wandelt diese um
				coordinates[k]= abs( indizies[ ListPos[k] ] ) - 1;
				if( indizies[ ListPos[k] ] < 0) {
					vz *= -1;
					}
				}
			sum += vz * werte[getPosition(coordinates)];
			Counters::counternext(current,shape,0);
			}
		return sum;
		}
	}
	
double IDC::getCellValue(const vector<int> &coordinates) {
	return getRangeSum(coordinates, coordinates);
	}

void IDC::ConcernedCellsUpdateRoutine(const vector<int> &index, int totalNumber, const vector<int> &shape, const vector<int> &posIndex, double dif) {
	//Hilfsarray, das dafuer sorgt, dass alle moeglichen Indexkombinationen durchlaufen werden.
	vector<int> current(dimensionality,0);
	//enthaelt die Positionen in der Liste, wo die Koordinaten drinstehen
	vector<int> listPos(0);
	//enthaelt die Koordinaten, die zur Summe dazugerechnet werden muessen.
	vector<int> coordinates(dimensionality,0);
	//Durchlaeuft alle betroffenen Zellen und updated diese
	for (int i=0; i<totalNumber; i++){
		listPos=getListPositions(current, posIndex);
		for (int k=0; k<dimensionality; k++){
			//liest Koordinaten aus, wandelt diese um
			coordinates[k]=index[ listPos[k] ];
		}
		//hier findet die tatsaechliche Aenderung statt. 
		werte[getPosition(coordinates)] += dif;
		//geht zur naechsten Koordinate
		Counters::counternext(current,shape,0);
		}
	}

// TODO: Durch arbeiten, es ist nicht sauber implementiert!
void IDC::OneCellUpdateDif(const vector<int> &current, double dif, double oldValue) {
	// Ermitteln der betroffenen Zellen
	vector<int> index(0);
	vector<int> posIndex(dimensionality, 0);
	vector<int> shapeLocal(dimensionality, 0);
	int totalNumber = 1;
	vector<int> tmp(0);
	//Fuegt alle betroffenen Indizes zu einer ArrayList zusammen
	for (int i=0; i<dimensionality; i++) {
		tmp= (dimensions[i].getStrategie())->getCellsForCellUpdate(current[i]);
		shapeLocal[i] = tmp.size();
		totalNumber *= shapeLocal[i];
		if (i > 0){
			posIndex[i] = posIndex[i-1] + shapeLocal[i-1];
			}
		index.insert(index.end(),tmp.begin(),tmp.end());
		}
	//Update der RealSize
	double newValue=oldValue + dif;
	// TODO: Problem mit double Genauigkeit beachten!
	if (oldValue != 0 && abs(oldValue) < 0.00000001) {
		oldValue = 0;
		}        
	if (newValue != 0 && abs(newValue) < 0.00000001) {
		newValue = 0;
		}

	if (newValue == 0) {
		if (oldValue != 0) // Wert wird auf 0 gesetzt
			{
			realSize--;
			}
		} else if ( oldValue == 0 ){
			realSize++;
		}
		//ruft Hilfsmethode auf, die alle betroffen Zellen aendert
	ConcernedCellsUpdateRoutine(index, totalNumber, shapeLocal, posIndex, dif);
	}

void IDC::OneCellUpdateFac(const vector<int> &current, double fac) {
	double oldValue=getCellValue(current);
	// alter Wert war 0
	if (abs(oldValue)<0.0001)
		return;
	//Differenz zwischen neuem Wert fac*oldValue und altem Wert wird berechnet
	double dif = abs(oldValue * (fac-1));
	OneCellUpdateDif(current, dif, oldValue);
	totalSum=totalSum+dif;
	}


// Komplet bearbeiten und verbessern. 
void IDC::rangeUpdate(const vector<int> &lower, const vector<int> &upper, double value, int strat) {
	if( !rangeValidator(lower,upper)) {
		return;
		}
	vector<int> upperBound(upper);
	for (unsigned int i=0; i<upperBound.size(); i++){
		upperBound[i]++;
		}
	// einfacher Fall, nur eine Zelle update

	if (lower == upper) {
		double oldValue = getCellValue(lower);
		if (oldValue == value){
			return;
			}
		double dif = value - oldValue;
		// Hilfsmethode wird aufgerufen
		OneCellUpdateDif(lower, dif, oldValue);
		totalSum=totalSum-oldValue+value;
		} else {
			double oldValue = getRangeSum(lower, upper);
			double dif = value - oldValue;
			if ( dif == 0) {
				return;
				}
			//Anzahl der betroffenen Zellen ermitteln
			int count = 1;
			for ( int i = 0; i < dimensionality; i++){
				count *= upper[i] - lower[i] + 1;
			}
			// neuer Wert ist 0
			if (value == 0 ) {
				vector<int> current(lower);
				//durchlaeuft alle betroffenen Zellen und setzt sie auf 0
				double currentValue = 0;
				for (int i = 0; i<count; i++) {
					currentValue = getCellValue(current);
					OneCellUpdateDif(current,(0-currentValue), currentValue);
					Counters::counternextLow(current, lower, upperBound);
					}
				totalSum=totalSum-oldValue;
				return;
			}
			if ( strat == 0){
				//Wieviel muss zu jeder Zelle hinzugefuegt werden
				dif = dif/count;
				vector<int> current(lower);
				//durchlaeuft alle betroffenen Zellen und erhoeht dabei jede Zelle mithilfe der Hilfsmethode um dif.
				double currentValue=0; 
				for ( int i = 0; i < count; i++){
					currentValue = getCellValue(current);
					OneCellUpdateDif(current,dif,currentValue);
					Counters::counternextLow(current, lower, upperBound);
				}
				totalSum=totalSum-oldValue+value;
			} 
			else{
				//alter Wert ist null. Keine Erhoehung um einen Faktor moeglich
				if(oldValue == 0) {
					rangeUpdate(lower, upper, value, 0);
					return;
				}
				// Um welchen Faktor wird erhoeht?
				double fac = value/oldValue;
				vector<int> current(lower);
				// alles muss auf null gesetzt werden
				if (fac==0){
					double currentValue = 0;
					for (int i=0; i<count; i++)	{
						currentValue = getCellValue(current);
						OneCellUpdateDif(current,(0-currentValue), currentValue);
						Counters::counternextLow(current, lower, upperBound);
					}
					totalSum=totalSum-oldValue+value;
					return;
				}
				//durchlaeuft alle betroffenen Zellen und erhoeht dabei jede Zelle mithilfe der Hilfsmethode um fac
				for(int i=0; i<count; i++){
					OneCellUpdateFac(current, fac);
					Counters::counternextLow(current, lower, upperBound);
					}
			}
		}
	}




	

File: src/structures/IDC.java
package structures;

import helpers.Counters;
import java.io.Serializable;
import java.util.ArrayList;

/**
 * Datenstruktur zum Speichern von Cubes.
 * @author Philippe Hagedorn und Zurab Khadikov
 * 
 */
public class IDC implements OlapCube, Serializable {

    /**
     * Eindimensionales Hilfsarray feur Werte
     */
    protected double werte[];
    protected int realSize;
    protected int dimensionality;
    public IDCDimension[] dimension;
    /**
     * Hilfsarray zur Bestimmung der Position im
     * eindimensionalen Array "werte"
     */
    private int[] positions;
    /**
     * Gibt an, ob Vorberechnung schon stattgefunden hat.
     */
    private boolean precomputed;

    /*
     * (non-Javadoc)
     * 
     * @see structures.OlapCube#initialize(int[], java.lang.String[])
     */
    public IDC() {
    }

    /**
     * Initialiesierungsmethode, dabei wird Hautstruktur von IDC Cube festgelegt
     * @param dimensions Dimensionssgoessen 
     * @param description Beschreiburng fuer jede Dimension
     */
    public void initialize(int[] dimensions, String[] description) {
        try {
            this.dimensionality = dimensions.length;
            this.positions = new int[dimensionality];
            this.dimension = new IDCDimension[dimensionality];

            //Setzen der Strategien sollte auch anders gehen koennen, jetzt Default Strategie
            for (int i = 0; i < dimensionality; i++) {
                this.dimension[i] = new IDCDimension(dimensions[i], description[i]);
            }

            int tmp = 1;
            this.positions[dimensionality - 1] = 1;
            for (int i = 2; i <= dimensionality; i++) {
                tmp *= dimensions[dimensionality - i + 1];
                this.positions[dimensionality - i] = tmp;
            }

            int size = tmp * dimensions[0];
            werte = new double[size];
            precomputed = false;

        } catch (Exception e) {
            System.err.println("IDC DataCube konnte nicht initialisiert werden!");
            System.err.println("Es gab folgende Fehler: " + e);
        }

    }

    /*
     * (non-Javadoc)
     * 
     * @see structures.OlapCube#setCellValue(int[], double)
     */
    public void setCellValue(int[] coordinates, double value) {
        if (coordinatesValidation(coordinates)) {
            int pos = getPosition(coordinates);
            if (!precomputed) {
                if (value == 0) {
                    if (this.werte[pos] != 0) // Wert wird auf 0 gesetzt
                    {
                        realSize--;
                    }
                } else if (this.werte[pos] == 0) {
                    realSize++;
                }
                this.werte[pos] = value;
            } else {
                rangeUpdate(coordinates, coordinates, value, 0);
            }
        }
    }

    /*
     * (non-Javadoc)
     * 
     * @see structures.OlapCube#getCellValue(int[])
     */
    public double getCellValue(int[] coordinates) {
        return this.getRangeSum(coordinates, coordinates);
    }

    /*
     * (non-Javadoc)
     * 
     * @see structures.OlapCube#getRangeSum(int[], int[])
     */
    public double getRangeSum(int[] lower, int[] upper) {
        if (!this.rangeValidator(lower, upper)) {
            return 0;
        }
        //Speichert das Zwischenergebnis
        double sum = 0;
        //Enthaelt die Indizes fuer alle Dimensionen
        ArrayList<Integer> Indizes = new ArrayList<Integer>();
        //pos[i] enthaelt die erste Position, die einen Index fuer die Dimension i enthaelt
        int[] pos = new int[this.dimensionality];
        //Hilfsvariable zur Berechnung des korrekten pos-Arrays
        int position = 0;
        //Speichert die Anzahl an Werten, die man zur Summe hinzufuegen muss
        int totalNumber = 1;
        //Fuegt die Indizes aller Dimensionen zu einer Liste zusammen
        for (int i = 0; i < this.dimensionality; i++) {
            ArrayList<Integer> tmp = this.dimension[i].getStrategie().getCellsForRangeSum(lower[i], upper[i]);
            Indizes.addAll(tmp);
            pos[i] = tmp.size();
            totalNumber *= pos[i];
            position += pos[i];
        }
        //Enthaelt fuer jede Dimension Anzahl der Indizes
        int[] shape = pos.clone();
        //Berechnet korrektes Indizes Array
        for (int i = this.dimensionality - 1; i >= 0; i--) {
            pos[i] = position - pos[i];
            position = pos[i];
        }
        //Hilfsarray, das dafuer sorgt, dass alle moeglichen Indexkombinationen durchlaufen werden.
        int[] current = new int[this.dimensionality];
        //enthaelt die Positionen in der Liste, wo die Koordinaten drinstehen
        int[] ListPos;

        for (int i = 0; i < totalNumber; i++) {
            ListPos = this.getListPositions(current, pos);
            //enthaelt die Koordinaten, die zur Summe dazugerechnet werden muessen.
            int[] coordinates = new int[this.dimensionality];
            //Vorzeichen des Summanden
            int vz = 1;
            for (int k = 0; k < this.dimensionality; k++) {
                //liest Koordinaten aus wandelt diese um
                coordinates[k] = Math.abs(Indizes.get(ListPos[k])) - 1;
                if (Indizes.get(ListPos[k]) < 0) {
                    vz *= -1;
                }

            }
            sum += vz * this.werte[this.getPosition(coordinates)];
            current = Counters.counternext(current, shape, 0);
        }

        return sum;
    }

    /**
     * Hilfsmethode, um Positionen von Indizes im Hilfsarray zu ermitteln.
     * @param coordinates Koordinaten bzgl. Hilfsarray
     * @param pos Positionsarray (wo faengt welche Dimension an)
     * @return Koordinate bzgl. Cube
     */
    protected int[] getListPositions(int[] coordinates, int[] pos) {
        int[] result = new int[coordinates.length];
        for (int i = 0; i < coordinates.length; i++) {
            result[i] = coordinates[i] + pos[i];
        }
        return result;
    }

    /*
     * (non-Javadoc)
     * 
     * @see structures.OlapCube#getShape()
     */
    public int[] getShape() {
        int[] dimensions = new int[dimensionality];
        for (int i = 0; i < dimensionality; i++) {
            dimensions[i] = this.dimension[i].getSize();
        }
        return dimensions;
    }

    /*
     * (non-Javadoc)
     * 
     * @see structures.OlapCube#getDescription()
     */
    public String[] getDescription() {
        String[] description = new String[dimensionality];
        for (int i = 0; i < dimensionality; i++) {
            description[i] = this.dimension[i].getName();
        }
        return description;
    }

    /*
     * (non-Javadoc)
     * 
     * @see structures.OlapCube#getSize()
     */
    public int getSize() {
        return realSize;
    }
    
    /**
     * Alternative Methode um Real Size zurück zu geben.
     * @return Tatsaechliche Anzahl von Zellen im Cube, die nicht gleich Null sind.
     */
    public int getSizeAlt() {
        int realSize=0;
        for(int i=0; i<this.werte.length; i++){
            if(this.werte[i]!=0)
                realSize++;
        }   
        return realSize;
    }

    /*
     * (non-Javadoc)
     * 
     * @see structures.OlapCube#rangeUpdate(int[] lower, int[] upper, double
     *      value,int strategieNum)
     */
    public void rangeUpdate(int[] lower, int[] upper, double value, int strat) {
        if (!this.rangeValidator(lower, upper)) {
            return;
        }
        int[] upperBound = upper.clone();
        for (int i = 0; i < upperBound.length; i++) {
            upperBound[i]++;
        }
        //einfacher Fall, nur eine Zelle updaten
        if (lower == upper) {
            double oldValue = this.getCellValue(lower);
            if (oldValue == value) {
                return;
            }
            double dif = value - oldValue;
            int pos = this.getPosition(lower);
            
            //Hilfsmethode wird aufgerufen
            CellUpdateDif(lower, dif);
        } else {
            double oldValue = this.getRangeSum(lower, upper);            
            double dif = value - oldValue;
            if (dif == 0) {
                return;
            }
            //Anzahl der betroffenen Zellen ermitteln
            int count = 1;
            for (int i = 0; i < this.dimensionality; i++) {
                count *= upper[i] - lower[i] + 1;
            }
            //neuer Wert ist 0
            if (value==0){
                int[] current=lower.clone();
                //durchlaeuft alle betroffenen Zellen und setzt sie auf 0
                for (int i = 0; i < count; i++) {
                    this.setCellValue(current, 0);
                    current = Counters.counternextLow(current, lower, upperBound);
                }
                return;
            }
            if (strat == 0) {
                //Wieviel muss zu jeder Zelle hinzugefuegt werden
                dif = dif / count;
                int[] current = lower.clone();
                //durchlaeuft alle betroffenen Zellen und erhoeht dabei jede Zelle mithilfe der Hilfsmethode um dif.
                for (int i = 0; i < count; i++) {
                    CellUpdateDif(current, dif);
                    current = Counters.counternextLow(current, lower, upperBound);
                }
            } else {
                //alter Wert ist null. Keine Erhoehung um einen Faktor moeglich
                if (oldValue == 0) {
                    rangeUpdate(lower, upper, value, 0);
                    return;
                }
                //Um welchen Faktor wird erhoeht?
                double fac = value / oldValue;
                int[] current = lower.clone();
                //alles muss auf null gesetzt werden.
                if (fac == 0) {
                    for (int i = 0; i < count; i++) {                        
                        this.setCellValue(current, 0);
                        current = Counters.counternextLow(current, lower, upperBound);
                    }
                    return;
                }
                //durchlaeuft alle betroffenen Zellen und erhoeht dabei jede Zelle mithilfe der Hilfsmethode um fac
                for (int i = 0; i < count; i++) {
                    CellUpdateFac(current, fac);
                    current = Counters.counternextLow(current, lower, upperBound);
                }
            }

        }

    }

    /**
     * Hilfsmethode zum Aendern einer einzelnen Zelle um einen Summanden.
     * @param current Koordinaten der Zelle, die aktualisiert werden soll.
     * @param dif Wert, um den erhoeht wird.
     */
    private void CellUpdateDif(int[] current, double dif) {
        // Ermitteln der betroffenen Zellen			
        ArrayList<Integer> Index = new ArrayList<Integer>();
        int[] positions = new int[this.dimensionality];
        int[] shape = new int[this.dimensionality];
        int totalNumber = 1;
        //Fuegt alle betroffenen Indizes zu einer ArrayList zusammen
        for (int i = 0; i < this.dimensionality; i++) {
            ArrayList<Integer> tmp = this.dimension[i].getStrategie().getCellsForCellUpdate(current[i]);
            shape[i] = tmp.size();
            totalNumber *= shape[i];
            if (i > 0) {
                positions[i] = positions[i - 1] + shape[i - 1];
            }
            Index.addAll(tmp);
        }
        //Update der RealSize
        double oldValue=this.getCellValue(current);
        double value=oldValue+dif;
        // Problem mit double Genauigkeit!
        if (oldValue != 0 && Math.abs(oldValue) < 0.0001) {
            oldValue = 0;
        }        
        if (value != 0 && Math.abs(value) < 0.0001) {
            value = 0;
        }
        if (value == 0) {
            if (oldValue != 0) // Wert wird auf 0 gesetzt
            {
                realSize--;
            }
        } else if (oldValue == 0) {
            realSize++;
        }
        //ruft Hilfsmethode auf, die alle betroffen Zellen aendert
        CellsUpdate(Index, totalNumber, shape, positions, dif);

    }

    /**
     * Hilfmethode zum Aendern einer einzelnen Zelle um einen Faktor
     * @param current Koordinaten der Zelle, die aktualisiert werden soll.
     * @param fac Faktor, um den erhoeht wird.
     */
    private void CellUpdateFac(int[] current, double fac) {
        double oldValue = this.getCellValue(current);
        // alter Wert war 0
        if (Math.abs(oldValue)<0.0001)
            return;
        //Differenz zwischen neuem Wert fac*oldValue und altem Wert wird berechnet
        double dif = oldValue * (fac - 1);
        CellUpdateDif(current, dif);
    }

    /**
     * Aendert durch ein Zell-Update betroffene Zellen 
     * @param Index Liste aller betroffenen Indizes
     * @param totalNumber Anzahl der betroffenen Indizes
     * @param shape Anzahl der Indizes in jeder Dimension
     * @param positions Erster Index jeder Dimension in der Liste
     * @param dif Wert, der addiert werden soll.
     */
    private void CellsUpdate(ArrayList<Integer> Index, int totalNumber, int[] shape, int[] positions, double dif) {
        //Hilfsarray, das dafuer sorgt, dass alle moeglichen Indexkombinationen durchlaufen werden.
        int[] current = new int[this.dimensionality];
        //enthaelt die Positionen in der Liste, wo die Koordinaten drinstehen
        int[] ListPos;
        //Durchlaeuft alle betroffenen Zellen und updated diese
        for (int i = 0; i < totalNumber; i++) {
            ListPos = this.getListPositions(current, positions);
            //enthaelt die Koordinaten, die zur Summe dazugerechnet werden muessen.
            int[] coordinates = new int[this.dimensionality];
            for (int k = 0; k < this.dimensionality; k++) {
                //liest Koordinaten aus wandelt diese um
                coordinates[k] = Index.get(ListPos[k]);
            }
            //hier findet die tatsaechliche Aenderung statt.         
            this.werte[this.getPosition(coordinates)] += dif;          
            //geht zur naechsten Koordinate
            current = Counters.counternext(current, shape, 0);
        }
    }

    /**
     * Checkt ob eine Range korrekt ist
     * @param lower Koordinaten der Untergrenze
     * @param upper Koordinaten der Obergrenze
     */
    public boolean rangeValidator(int[] lower, int[] upper) {
        boolean valid = true;
        for (int i = 0; i < lower.length; i++) {
            if (lower[i] - upper[i] > 0) {
                System.err.println("Rangekoordinaten sind falsh geweahlt!");
                valid = false;
            }
        }
        return valid;
    }

    /**
     * Die Methode laesst fuer jede Dimension die Vorberechnungsstrategie anwenden.
     * Es wird das Visitor Pattern verwendet.
     */
    public void precompute() {
        if (this.precomputed) {
            return;
        }
        for (int i = 0; i < this.dimensionality; i++) {
            IStrategie tmp = dimension[i].getStrategie();
            //Standardstrategie setzen
            if (tmp == null) {
                dimension[i].setStrategie(new LPS(dimension[i].size, (int) Math.sqrt(dimension[i].size)));
            }
            dimension[i].getStrategie().precompute(this, i);

        }
        this.precomputed = true;
    }

    /**
     * Hilfsmethode, die umrechnet eingegebene Koordinaten auf die Position im
     * eindimensionalen Wertearray.
     * 
     * @param coordinates mehrdimensionale Koordinate
     * @return Position im Werte Array
     */
    public int getPosition(int[] coordinates) {
        int pos = 0;
        for (int i = 0; i < coordinates.length; i++) {
            pos += coordinates[i] * this.positions[i];
        }
        return pos;
    }

    /**
     * Diese Methode prueft ob die Koordinaten korrekt eingegeben sind.
     * 
     * @param coordinates Koordinate
     * @return "true" wenn ales in Ordnung ist
     */
    public boolean coordinatesValidation(int[] coordinates) {
        boolean valid = true;
        if (this.dimensionality == 0) {
            valid = false;
            System.err.println("DataCube ist nicht initialisiert!");
        } else {
            if (coordinates.length != this.dimensionality) {
                System.err.println("Koordinaten Eingabe ist ungueltig!");
                valid = false;// ungueltige Eingabe
            } else {
                int[] upper = getShape();
                for (int i = 0; i < coordinates.length; i++) {
                    if (coordinates[i] < 0 || coordinates[i] >= upper[i]) {
                        System.err.println("Koordinaten sind ausserhalb des Dimensions-Bereichs!");
                        valid = false;
                    }
                }
            }
        }
        return valid;
    }
}
